### ISSUE TITLE
Handling of newlines when table elements contain elements like: p and div

### ISSUE THREAD

**Issue by janis-veinbergs** [mysticmind/reversemarkdown-net#63]
My goal is to have more human readable tables. Is this actually the goal of reversemarkdown library? Am I using it wrong?

Consider two input HTML that produce different outputs, but should produce identical:
```html
<html><body><table><tbody><tr><td><p>col1</p></td><td><p>col2</p></td></tr><tr><td><p>data1</p></td><td><p>data2</p></td></tr></tbody></table></body></html>
```
Out: 
```
| col1<br> | col2<br> |
| --- | --- |
| data1<br> | data2<br> |
```

```html
<html><body><table><tbody><tr><td><p>
col1</p></td><td><p>col2</p></td></tr><tr><td>
<p>
data1
</p>
</td>
<td><p>data2</p></td></tr></tbody></table></body></html>
```
Out:
```
| col1<br> | col2<br> |
| --- | --- |
| <br>data1<br> | data2<br> |
```
I would expect output in both cases to be this:
```
| col1 | col2 |
| --- | --- |
| data1 | data2 |
```

There are 2 issues
# `p` converter no matter what appends newline to end: 
https://github.com/mysticmind/reversemarkdown-net/blob/2a29c36c5ade6534fb5fdb99f1aea53b8a4ac5c2/src/ReverseMarkdown/Converters/P.cs#L18

Browser doesn't. And Outlook generates tables just like that - `td > p > #text`

The incomplete fix would be checking if `p` (perhaps any flow content) is last element within `td` and not add trailing newline. 

So considering some scenarios:
1. `<td><p>data1</p></td>`: browser renders no newlines. Reversemarkdown excess ending br: `| data1<br> |`
2. `<td>data1<p>p</p></td>`. Browser renders newline before p. Reversemarkdown excess ending <br>: `| data1<br>p<br> |`
3. `<td><div><p>data1</p></div></td>`, browser renders no newlines. Reversemarkdown excess starting and ending br: `| <br>data1<br> |`

I don't know what would be the best way to handle these newlines. Because when I convert real-life html that comes from outlook, it is just overwhelmed with newlines. 

Cases are many. Fixing those I brought up is possible. I should probably do it? 

# line break after starting tag and before ending tag should be ignored 
This is some non-standard document, but currently browser behaves like that:
> A line break occurring immediately following a start tag must be ignored, as must a line break occurring immediately before an end tag. This applies to all HTML elements without exceptions. In addition, for all elements except `PRE`, leading white space characters, such as spaces, horizontal tabs, form feeds and line breaks, following the start tag must be ignored, and any subsequent sequence of contiguous white space characters must be replaced by a single word space.
> The following three examples must be rendered identically:
```html
<P>Thomas is watching TV.</P>
```
```html
<P>
Thomas is watching TV.
</P>
```
```html
<P>
   Thomas is watching TV.
</P>
```
[w3.org HTML Text element](https://www.w3.org/TR/WD-html40-970917/struct/text.html)

**Comment by mysticmind** [comment]
@janis-veinbergs when you say "human readable tables", are you referring to alignment of the column so that it is not jagged?

With regards to the use cases with p tags within td converting to `br` tags, will check and get back to you.

**Comment by janis-veinbergs** [comment]
> are you referring to alignment of the column so that it is not jagged?

@mysticmind Input on the left, output on the right: https://i.imgur.com/PfrpFhR.png or input being:

```html
<table class="x_MsoTableGrid" border="1" cellspacing="0" cellpadding="0" style="color: rgb(33, 33, 33); font-family: wf_segoe-ui_normal, &quot;Segoe UI&quot;, &quot;Segoe WP&quot;, Tahoma, Arial, sans-serif, serif, EmojiFont; font-size: 15px; background-color: rgb(255, 255, 255); border-collapse: collapse; border: none;">
<tbody>
<tr>
<td width="288" valign="top" style="width: 215.75pt; border: 1pt solid windowtext; padding: 0cm 5.4pt;">
<p class="x_MsoNormal" style="margin: 0cm 0cm 0.0001pt; font-size: 11pt; font-family: Calibri, sans-serif;">
a</p>
</td>
<td width="288" valign="top" style="width: 215.75pt; border-top: 1pt solid windowtext; border-right: 1pt solid windowtext; border-bottom: 1pt solid windowtext; border-image: initial; border-left: none; padding: 0cm 5.4pt;">
<p class="x_MsoNormal" style="margin: 0cm 0cm 0.0001pt; font-size: 11pt; font-family: Calibri, sans-serif;">
b
</p>
</td>
</tr>
<tr>
<td width="288" valign="top" style="width: 215.75pt; border-right: 1pt solid windowtext; border-bottom: 1pt solid windowtext; border-left: 1pt solid windowtext; border-image: initial; border-top: none; padding: 0cm 5.4pt;">
<p class="x_MsoNormal" style="margin: 0cm 0cm 0.0001pt; font-size: 11pt; font-family: Calibri, sans-serif;">
1</p>
</td>
<td width="288" valign="top" style="width: 215.75pt; border-top: none; border-left: none; border-bottom: 1pt solid windowtext; border-right: 1pt solid windowtext; padding: 0cm 5.4pt;">
<p class="x_MsoNormal" style="margin: 0cm 0cm 0.0001pt; font-size: 11pt; font-family: Calibri, sans-serif;">
2</p>
</td>
</tr>
</tbody>
</table>
```
Output:
```
| 
a
| 
b
|
| --- | --- |
| 
1
| 
2
|
```

I think that https://github.com/mysticmind/reversemarkdown-net/blob/2a29c36c5ade6534fb5fdb99f1aea53b8a4ac5c2/src/ReverseMarkdown/Converters/P.cs#L15-L19 could be rewritten as:

```csharp
        public override string Convert(HtmlNode node)
        {
            var indentation = IndentationFor(node);
            var newlineAfter = NewlineAfter(node);

            return $"{indentation}{TreatChildren(node).Trim()}{newlineAfter}";
        }

        private string NewlineAfter(HtmlNode node) {
            var lastParagraphInTd = node.NextSibling == null && (node.ParentNode.Name == "td" || node.ParentNode.Name == "th");
            return lastParagraphInTd ? "" : Environment.NewLine;
        }
```

Perhaps more cases can be added - if `p` is last within `div`, `blockquote` node then 2 newlines would be added. Maybe they are cleaned up somwhere, don't know. Maybe it is easyer to somwhere reduce 3+ newlines to 2, because well, markdown doesn't actually care if there are 2 or 10 newlines.

**Comment by mysticmind** [comment]
@janis-veinbergs proposed change pertaining to handling of paragraph within `td` looks fine to me.

**Comment by janis-veinbergs** [comment]
@mysticmind  see Pull Request #64. 
The result is more complex as `p` and `div` tags are handled within table.
Moreover I'v put in some logic that handles markup formatting: newlines, tabs. That is, consider `<td>\r\n\t<p></p></td> - the first Node would actually be `#text`. So I check for that too.

The end result is more readable tables, less linebreaks.

I ran it through previous example (table html Outlook generates) and got nicer table:
```
| a | b |
| --- | --- |
| 1 | 2 |
```

**Comment by mysticmind** [comment]
Fixed via #64

**Comment by mysticmind** [comment]
Available in v3.11.0 published few mins ago
